### PR TITLE
Run tests on Helix

### LINKED ISSUE #49605: Roslyn tests should run on Helix
https://github.com/dotnet/arcade/blob/67035ca98fc8a4395bf7900e9288d99ecd93a3f4/Documentation/AzureDevOps/SendingJobsToHelix.md

/cc @jaredpar @chsienki

### PR DESCRIPTION
**Pull request by RikkiGibson**
**Note**: this PR does not make us run on Helix in production but rather prepares us to "flip the bit" so that it's easy for us to move back to single-machine test jobs in case problems occur.

Closes #49605 

Overall this seems to improve CI run times by a good amount. We will need to put it into production before we really know the effect. The times you’re used to seeing are *underestimates* because they don’t include the provisioning time. The new times do include that, since the jobs are run on “thin” AzDo agents which provision quickly and then wait for Helix to provision+run the job.

We have had concerns about the perf of the Spanish queue—here it still takes quite a while. It seems like we should just try putting it in prod, and seeing how the queues scale and serve us over time, keeping in mind that we are the first/only users of this queue.

Here is a failing run so you can see that failing tests are published properly: https://dev.azure.com/dnceng/public/_build/results?buildId=937555&view=results

### CODE REVIEW

File: src/Tools/Source/RunTests/TestRunner.cs [hunk 1]
```diff
@@ -40,6 +42,95 @@ internal TestRunner(Options options, ProcessTestExecutor testExecutor)
             _options = options;
         }
 
+        internal async Task<RunAllResult> RunAllOnHelixAsync(IEnumerable<AssemblyInfo> assemblyInfoList, CancellationToken cancellationToken)
+        {
+            var sourceBranch = Environment.GetEnvironmentVariable("BUILD_SOURCEBRANCH");
+            if (sourceBranch is null)
+            {
+                ConsoleUtil.WriteLine("BUILD_SOURCEBRANCH is not set");
+                sourceBranch = "local";
```
Comment (jaredpar) [inline src/Tools/Source/RunTests/TestRunner.cs:51]:
```suggestion
                sourceBranch = "local";
                ConsoleUtil.WriteLine("BUILD_SOURCEBRANCH is not set using {sourceBranch}");
```

File: src/Tools/Source/RunTests/TestRunner.cs [hunk 2]
```diff
@@ -40,6 +42,95 @@ internal TestRunner(Options options, ProcessTestExecutor testExecutor)
...
+            File.WriteAllText("helix-tmp.csproj", project);
+            var process = ProcessRunner.CreateProcess(
+                executable: _options.DotnetFilePath,
+                arguments: "build helix-tmp.csproj",
+                captureOutput: true,
+                onOutputDataReceived: (e) => ConsoleUtil.WriteLine(e.Data),
+                cancellationToken: cancellationToken);
+            var result = await process.Result;
+
+            // TODO: how do we handle publishing stuff like proc dumps when test runs have crashes?
```
Comment (jaredpar) [inline src/Tools/Source/RunTests/TestRunner.cs:103]: Have we asked core-eng how this should be handled?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -44,8 +44,7 @@ jobs:
     buildJobName: Build_Windows_Debug
     testArtifactName: Transport_Artifacts_Windows_Debug
     configuration: Debug
-    testArguments: -testDesktop -test32
-    queueName: 'BuildPool.Windows.10.Amd64.ES.VS2017.Open'
+    testArguments: -testDesktop -test32 -helixQueueName Windows.10.Amd64.Server19H1.ES.Open
```
Comment (jaredpar) [inline azure-pipelines.yml:47]: There will be some hiccups when we migrate to Helix, complex system stuff will go wrong. I feel like it's invetible we'll hit a few periods where we want to flip away from Helix, go back to the old system while we work out the kinks and then flip back. 

Ideally there would just be a nice global variable we flip in the YML to achieve this. That though feels like it will make the scripts really awkward: double the queue information, lot more `if` in them to change how it works. Definitely would be hard to follow and a bit more up front work. 

At the same time the only other option for "temporarily disable Helix" is revert this change. This change is more than just flipping our infra to Helix, it's also changing how run tests work, changing how our scripts work, etc ... That could be done independently because they're all designed to work on Helix or without Helix. Wonder if we should do the following: 

1. Get this change to a point where we all agree and approve it. 
1. Do not commit this change 
1. Open two new PRs: one PR that contains everything but the YML changes and one that contains only the YML changes. Merge them in that order. 

Then if we do hit a bug where we need to temporarily disable Helix we would just be reverting the YML. That's probably more managable than all the other work that is coming along here. 

Thoughts? Basically typing while I'm thinking on this one.

File: .vscode/launch.json [hunk 4]
```diff
@@ -11,7 +11,7 @@
       "preLaunchTask": "build",
       // If you have changed target frameworks, make sure to update the program path.
       "program": "${workspaceFolder}/artifacts/bin/RunTests/Debug/netcoreapp3.1/RunTests.dll",
-      "args": ["--tfm", "netcoreapp3.1", "--sequential", "--html"],
+      "args": ["--tfm", "net5.0", "--helix", "--html"],
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply .vscode/launch.json:14]: It was convenient for development, but I'll revert it to simplify the change.

File: eng/pipelines/test-unix-job.yml [hunk 5]
```diff
@@ -44,31 +33,24 @@ jobs:
...
-    - task: PublishTestResults@2
-      displayName: Publish xUnit Test Results
-      inputs:
-        testRunner: XUnit
-        testResultsFiles: '$(Build.SourcesDirectory)/artifacts/TestResults/${{ parameters.configuration }}/*.xml'
-        mergeTestResults: true
-        testRunTitle: ${{ parameters.testRunName }}
-      condition: always()
+      env:
+        SYSTEM_ACCESSTOKEN: $(System.AccessToken)
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply eng/pipelines/test-unix-job.yml:53]: We need to send this token over to the helix machine for it to be able to publish test results. This token is usually available in the pipeline yml but not in the environment unless specifically included.

File: src/Tools/Source/RunTests/TestRunner.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Tools/Source/RunTests/TestRunner.cs:103]: It sounds like Helix machines automatically set up the configuration for generating dumps when the test process crashes. I'll just play with adding crashes to the tests and see if I can get them to publish and see if we need to do anything else.

File: eng/build.ps1 [hunk 6]
```diff
@@ -178,6 +180,11 @@ function Process-Arguments() {
     exit 1
   }
 
+  if ($testVsi -and $helix) {
+    Write-Host "Cannot run integration tests on Helix"
+    exit 1
+  }
+
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply eng/build.ps1:187]: I would be happy to make this change if we actually push the decision on the helixQueueName into the scripts--currently if you don't specify a helixQueueName then RunTests will just default it to the Windows 10 queue.

File: global.json [hunk 7]
```diff
@@ -12,6 +12,7 @@
     "xcopy-msbuild": "16.8.0-preview2.1"
   },
   "msbuild-sdks": {
-    "Microsoft.DotNet.Arcade.Sdk": "1.0.0-beta.20621.2"
+    "Microsoft.DotNet.Arcade.Sdk": "1.0.0-beta.20621.2",
+    "Microsoft.DotNet.Helix.Sdk": "6.0.0-beta.20573.2"
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply global.json:16]: If we configure this properly then the dotnet-maestro will update it the same way it updates the rest of the arcade dependencies. For example dotnet/runtime#44020

[the remark this answers is not in the document]
Comment (RikkiGibson) [reply global.json:16]: It does seem, though, that we should require the same major version of both sdks. I'll have to experiment and follow up on that.

File: azure-pipelines.yml (hunk 3 above)
Comment (RikkiGibson) [reply azure-pipelines.yml:47]: I think that to start with, a two-PR solution makes plenty of sense. If we want to disable helix in the short term then reverting the second PR will be a good way to do it. I also think that in the medium term we might end up wanting to revert helix on specific jobs, etc., or just "turn off" helix test runs after a number of intermediate changes have been merged into this file, at which point we'd probably want to create a "helix" pipeline variable.

Comment (RikkiGibson) [reply azure-pipelines.yml:47]: Note that *some* yaml changes will be needed to make CI work even with helix disabled. For example, we might need to introduce a command that searches for and runs all the `rehydrate.cmd/sh` scripts, or write out a top-level script that runs all of those scripts.

File: src/Tools/Source/RunTests/TestRunner.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Tools/Source/RunTests/TestRunner.cs]: Fixed

File: eng/pipelines/test-unix-job.yml [hunk 8]
```diff
@@ -17,23 +17,12 @@ parameters:
...
 - job: ${{ parameters.jobName }}
   dependsOn: ${{ parameters.buildJobName }}
   pool:
-    ${{ if ne(parameters.queueName, '') }}:
-      name: NetCorePublic-Pool
-      queue: ${{ parameters.queueName }}
-
-    ${{ if ne(parameters.vmImageName, '') }}:
-      vmImage: ${{ parameters.vmImageName }}
+    vmImage: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply eng/pipelines/test-unix-job.yml]: Done

File: eng/pipelines/test-windows-job.yml [hunk 9]
```diff
@@ -18,16 +18,12 @@ parameters:
...
-  type: string
-  default: 'BuildPool.Windows.10.Amd64.Open'
 
 jobs:
 - job: ${{ parameters.jobName }}
   dependsOn: ${{ parameters.buildJobName }}
   pool:
-    name: NetCorePublic-Pool
-    queue: ${{ parameters.queueName }}
+    vmImage: windows-latest
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply eng/pipelines/test-windows-job.yml]: Done

File: eng/build.ps1 [hunk 10]
```diff
@@ -395,6 +402,14 @@ function TestUsingRunTests() {
     $args += " --sequential"
   }
 
+  if (-not $testVsi -and ($helix -or $ci)) {
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply eng/build.ps1]: Fixed

File: src/Tools/Source/RunTests/TestRunner.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Tools/Source/RunTests/TestRunner.cs:116]: Here's an example PR which publishes dumps: https://dev.azure.com/dnceng/public/_build/results?buildId=949005&view=ms.vss-test-web.build-test-results-tab&runId=29928386&paneView=debug&resultId=210588

Note that in subsequent commits I did actually add the name of the job to the helix run name, so that the test failures list is more comprehensible.

Comment (jaredpar) [reply src/Tools/Source/RunTests/TestRunner.cs]: Nice. 

Still though I think we need to track down the mechanism by which this is occurring. There are a couple of different ways you can go about this and it would impact the situations in which we do or do not get dumps. Want to make sure they're doing this in a way that will get all the dumps we are accustomed to.

**Comment by RikkiGibson** [comment]
This PR is now a *preparation* for running tests on Helix. All the underlying stuff is changed except for .yml, and tests continue to run on the build agents.

In order to actually start running the tests on helix, we just have to revert the commit 9a69cec2544669d959b95f3ff3de9130d9cd7d3f

File: eng/Version.Details.xml [hunk 11]
```diff
@@ -11,5 +11,9 @@
       <Uri>https://github.com/dotnet/roslyn</Uri>
       <Sha>c34bb049f1f74256f109794509cd9f0f1c23faea</Sha>
     </Dependency>
+    <Dependency Name="Microsoft.DotNet.Helix.Sdk" Version="2.0.0-beta.21063.4">
```
Comment (RikkiGibson) [inline eng/Version.Details.xml:14]: note that this is the version given by just grabbing from the same DARC channel that we get the Arcade SDK from.

**Review by JoeRobich — APPROVED** [review]

**Review by jaredpar — APPROVED** [review]
